### DISCUSSION TITLE
macOS: `SIGKILL (Code Signature Invalid)` when running `/Applications/kitty.app/Contents/MacOS/kitty`

### DISCUSSION THREAD

**Discussion by huyz** [kovidgoyal/kitty#7576] [Q&A] [answered]
For me, running `open -a kitty.app` works just fine.  But for the past several months, every time I run `/Applications/kitty.app/Contents/MacOS/kitty` at the command line, I get an immediate crash:

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Incident Identifier: 05C6123B-99A3-4AA2-AE9F-395FB8176AB5
CrashReporter Key:   25736505-D900-4C30-9B4B-4DE7B97B7E0F
Hardware Model:      MacBookPro18,2
Process:             kitty [88242]
Path:                /Applications/kitty.app/Contents/MacOS/kitty
Identifier:          net.kovidgoyal.kitty
Version:             0.35.2 (0.35.2)
Code Type:           ARM-64 (Native)
Role:                Unspecified
Parent Process:      zsh [19675]
Coalition:           com.googlecode.iterm2 [1947]
Responsible Process: iTerm2 [19595]

Date/Time:           2024-06-27 02:32:00.2715 -0700
Launch Time:         2024-06-27 02:32:00.2699 -0700
OS Version:          macOS 14.5 (23F79)
Release Type:        User
Report Version:      104

Exception Type:  EXC_CRASH (SIGKILL (Code Signature Invalid))
Exception Codes: 0x0000000000000000, 0x0000000000000000
Termination Reason: CODESIGNING 1 Taskgated Invalid Signature

Triggered by Thread:  0

Thread 0 Crashed:
0                                 	       0x102690b70 _dyld_start + 0
```

Anyone have any idea what's going on?

I'm on: macOS 14.5 23F79 (Sonoma) MacBookPro18,2 (Apple M1 Max, arm64)

**Comment by kovidgoyal** [comment]
You have two different kitty's on your system, one of which has an
invalid code signing signature.

**Reply by huyz** [reply]
Thanks for the reply, Kovid.

I used Find Any File and the only other kitty I could find was in the Trash. Curiously, permanently deleting seems to have fixed the issue! (I'll only be sure after a week that the issue isn't coming back for some reason).

That's one mystery: why the contents of the Trash would matter.

The other mystery is why any kitty's would have an invalid code signing signature.  I've both done `brew install kitty` or installed from GitHub releases. And either way, the kitty executable works at first, but later (maybe hours later or the next day), I get the SIGKILL. Is it possible some other app or process of mine is corrupting the signature somehow?  I don't know anything about macOS development so I have no idea how this could even happen

**Reply by huyz** [reply]
Well, the error came back again. And I don't know why.

If I run [apparency](https://www.mothersruin.com/software/Apparency/) on the `/Applications/kitty.app/Contents/MacOS/kitty` binary, I don't see anything out of the ordinary:

![screenshot 2024-06-30T173647Z@2x](https://github.com/kovidgoyal/kitty/assets/128394/0984e1e3-19c7-444e-95e5-4863eccf5946)


Ok, well, the only idea I can come up with is to try ad-hoc signing:

```bash
codesign --force --sign - /Applications/kitty.app/Contents/MacOS/kitty
```

> If identity is the single letter "-" (dash), ad-hoc signing is performed.  Ad-hoc signing does not use an identity at all, and identifies exactly one instance of code. Significant restrictions apply to the use of ad-hoc signed code; consult documentation before using this.

🤞

**Reply by huyz** [reply]
Well nothing I do has helped, except for reinstalling and enjoying a few minutes to a few hours of crash-free use.

@kovidgoyal do you think this could be applicable: https://stackoverflow.com/a/77875189/161972

> I've had related problems to this, in my case it was because I was using an old version of macOS and Xcode, that didn't have the newer Apple Worldwide Developer Relations certificate installed.
> You can install it yourself by downloading it from https://www.apple.com/certificateauthority/.

**Comment by kovidgoyal** [comment]
Signatures are simply files stored inside the app bundle. Any
other software with write access to the bundle contents (which is any
software running as your user account) can corrupt them.

**Comment by kovidgoyal** [comment]
No clue, no harm in trying it. Although easier to just disable
gatekeeper and see if that solves your issues. It is total security
theatre anyway.

**Reply by huyz** [reply, marked as answer]
Well, running `sudo spctl --global-disable` didn't help, but building from source is a good enough workaround until I can figure this out some day.
